engines/lab/utils.cpp

```diff
@@ -0,0 +1,272 @@
+			sourceFile->read(dest, copy);
+			dest += copy;
+		} else {
+			while (copy) {
+				*dest = sourceFile->readUint16LE();
+				dest++;
+				copy--;
+			}
+		}
+	}
+}
+
+template<typename T>
+void Utils::verticalUnDiff(T *dest, Common::File *sourceFile, uint16 bytesPerRow) {
+	uint16 counter = 0;
+	byte bytesPerWord = sizeof(T);
+	uint16 wordsPerRow = bytesPerRow / bytesPerWord;
+
+	while (counter < wordsPerRow) {
+		T *curPtr = dest + counter;
+
+		for (;;) {
+			uint16 skip = sourceFile->readByte();
+			uint16 copy = sourceFile->readByte();
+
+			if (skip == 255) {
+				counter += copy;
+				break;
+			} else {
+				curPtr += (skip * wordsPerRow);
+
+				while (copy) {
+					if (bytesPerWord == 1)
+						*curPtr++ = sourceFile->readByte();
+					else if (bytesPerWord == 2)
+						*curPtr = sourceFile->readUint16LE();
+					else if (bytesPerWord == 4)
+						*curPtr = sourceFile->readUint32LE();
+					else
+						error("verticalUnDiff: Invalid bytesPerWord (%d)", bytesPerWord);
+					curPtr += wordsPerRow;
+					copy--;
+				}
+			}
+		}
+	}
+}
+
+void Utils::runLengthDecode(byte *dest, Common::File *sourceFile) {
+	int8 num;
+	int16 count;
+
+	while (1) {
+		num = sourceFile->readSByte();
+
+		if (num == 127) {
+			return;
+		} else if (num > '\0') {
+			sourceFile->read(dest, num);
+			dest   += num;
+		} else {
+			count = (int16)(-num);
+			num = sourceFile->readSByte();
+
+			while (count) {
+				*dest = num;
+				dest++;
+				count--;
+			}
+		}
+	}
+}
+
+void Utils::verticalRunLengthDecode(byte *dest, Common::File *sourceFile, uint16 bytesPerRow) {
+	int16 count;
+	byte *top = dest;
+
+	for (uint16 i = 0; i < _dataBytesPerRow; i++) {
+		dest = top;
+		dest += i;
+
+		int8 num = sourceFile->readSByte();
+
+		while (num != 127) {
+			if (num > '\0') {
+				while (num) {
+					*dest = sourceFile->readByte();
+					dest += bytesPerRow;
+					num--;
+				}
+			} else {
+				count = (int16)(-num);
+				num = sourceFile->readSByte();
+
+				while (count) {
+					*dest = num;
+					dest += bytesPerRow;
+					count--;
+				}
+			}
+
+			num = sourceFile->readSByte();
+		}
+	}
+}
+
+void Utils::unDiff(byte *newBuf, byte *oldBuf, Common::File *sourceFile, uint16 bytesPerRow, bool isVertical) {
+	sourceFile->skip(1);
+	byte bufType = sourceFile->readByte();
+
+	if (isVertical) {
+		if (bufType == 0)
+			verticalUnDiff<byte>(newBuf, sourceFile, bytesPerRow);
+		else if (bufType == 1)
+			verticalUnDiff<uint16>((uint16 *)newBuf, sourceFile, bytesPerRow);
+		else if (bufType == 3)
+			verticalUnDiff<uint32>((uint32 *)newBuf, sourceFile, bytesPerRow);
+		else
+			error("Unexpected variable compression scheme %d", bufType);
+	} else {
+		if (bufType == 0)
+			unDiff<byte>(newBuf, sourceFile);
+		else if (bufType == 1)
+			unDiff<uint16>((uint16 *)newBuf, sourceFile);
+		else
+			error("Unexpected compression scheme %d", bufType);
+	}
+}
+
+void Utils::setBytesPerRow(int num) {
+	_dataBytesPerRow = num;
+}
+
+uint16 Utils::getRandom(uint16 max) {
```

Comment:
If I understand this function correctly `max` isn't actually the maximum value this function shall return but rather the maximum value +1. Otherwise the call to our random functionality seems wrong and the if is completely confusing to me.